### PR TITLE
[LLVM] add LLVMMLIRUndef

### PR DESCRIPTION
**Pull request by webmiche**
This PR adds the [LLVMMLIRUNDEF(https://mlir.llvm.org/docs/Dialects/LLVM/#llvmmlirundef-mlirllvmundefop) operation to the llvm mirror.

`llvm.mlir.undef` allows to define a undefined value of any mlir type. In the lowering, mlir decided where to allocate and so on.

### CODE REVIEW

File: src/xdsl/dialects/llvm.py [hunk 1]
```diff
@@ -4,8 +4,10 @@
 from xdsl.irdl import (ParameterDef, AnyAttr, irdl_op_builder,
                        irdl_attr_definition, AttributeDef, OperandDef,
                        ResultDef, irdl_op_definition, builder)
-from xdsl.ir import MLContext, ParametrizedAttribute, TYPE_CHECKING, Attribute, Operation
-from xdsl.dialects.builtin import StringAttr, ArrayOfConstraint, ArrayAttr
+from xdsl.ir import (MLContext, ParametrizedAttribute, TYPE_CHECKING,
```
Comment (math-fehr) [inline src/xdsl/dialects/llvm.py:7]: My bad, I never realized that `TYPE_CHECKING` was imported from `xdsl.ir` previously.
Could you import it from `typing` instead?

File: src/xdsl/dialects/llvm.py [hunk 2]
```diff
@@ -61,3 +75,25 @@ class LLVMInsertValue(Operation):
...
+    res = ResultDef(LLVMPointerType)
+
+
+@irdl_op_definition
+class LLVMLoad(Operation):
+    name = "llvm.load"
+
+    addr = OperandDef(LLVMPointerType)
+
+    # TODO: there are several attributes missing here atm
```
Comment (math-fehr) [inline src/xdsl/dialects/llvm.py:97]:
```suggestion
    # TODO: there are several attributes missing here
```
;)

**Review by math-fehr — APPROVED** [review]
So, I'm not that familiar with pointers in LLVM, so I'm not sure if the test is actually correct.
Otherwise, looks good to me, I just added some really nit comments.

File: src/xdsl/mlir_converter.py [hunk 3]
```diff
@@ -67,6 +67,14 @@ def convert_type(self, typ: Attribute) -> mlir.ir.Type:
...
             return mlir_module.body.operations.__getitem__(
                 0).operation.result.type
+        if isinstance(typ, LLVMPointerType):
+            # TODO: This always gives an opaque ptr, even though we know the
+            # pointer type in xdsl.
+            mlir_module = mlir.ir.Module.parse("""
+                %0 = llvm.mlir.constant(1 : i32) : i32
+                %1 = llvm.alloca %0 x i8 : (i32) -> !llvm.ptr
+               """)
+            return mlir_module.body.operations.__getitem__(1).result.type
```
[the remark this answers is not in the document]
Comment (webmiche) [reply src/xdsl/mlir_converter.py:77]: There are no bindings for the LLVM dialect available at all, i.e. no ops, types or attributes. Operations work through their generic form tough.

[the remark this answers is not in the document]
Comment (Dinistro) [reply src/xdsl/mlir_converter.py:77]: Ah, this is what you already told me. 
May I ask why you even produce LLVMIR in xDSL and not stay at the MLIR dialect level?

**Comment by webmiche** [comment]
After some discussion with @ingomueller-net, I realised that I do not need allocas, loads and pointer types, but rather `llvm.mlir.undef` (at least that how  he does it in his bakcend). So I rewrote this PR to reflect that.

**Comment by webmiche** [comment]
@math-fehr this needs a rereview, as I have changed it fundamentally :)

**Comment by math-fehr** [comment]
Yeah, everything looks good to me as it is! If this is ready to merge, feel free to merge it!

**Comment by webmiche** [comment]
@math-fehr, @georgebisbas I had to restructure the imports of this PR during rebasing as the earlier PR broke the mlir-converter (the import order in the MLIRConverter actually matters). I think this is a problem that we need to address at osme point, but I do not see a solution. Can I still merge this? Or should we fix this problem first?

**Comment by georgebisbas** [comment]
You say the earlier PR broke the MLIR converter.
Does that mean there was no test to catch this problem?
If s, is that tested now?

It is OK with me to merge.
I think that slowly slowly all the imports goona be fixed while developing, as you will not have the "freedom", to import
any way you want.
So, we are going fine.

File: src/xdsl/dialects/llvm.py [hunk 4]
```diff
@@ -1,11 +1,13 @@
 from __future__ import annotations
 from dataclasses import dataclass
+from typing import TYPE_CHECKING
```
Comment (georgebisbas) [inline src/xdsl/dialects/llvm.py:3]: This is actually correct fix

**Review by georgebisbas — APPROVED** [review]

File: src/xdsl/dialects/llvm.py [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
 
 from xdsl.irdl import (ParameterDef, AnyAttr, irdl_op_builder,
                        irdl_attr_definition, AttributeDef, OperandDef,
                        ResultDef, irdl_op_definition, builder)
-from xdsl.ir import MLContext, ParametrizedAttribute, TYPE_CHECKING, Attribute, Operation
-from xdsl.dialects.builtin import StringAttr, ArrayOfConstraint, ArrayAttr
+from xdsl.ir import (MLContext, ParametrizedAttribute, Attribute, Operation)
+from xdsl.dialects.builtin import (StringAttr, ArrayOfConstraint, ArrayAttr,
+                                   IntegerAttr, IntegerType)
```
Comment (ingomueller-net) [inline src/xdsl/dialects/llvm.py:10]: @webmiche: Are you implementing Stella's suggestion from the other PR, here? I think her suggestion was to use parentheses *and have one package per line*. This makes reading the list itself easier, but also it means that adding and removing a package corresponds to adding and removing exactly one line, so the git diff is easier to read, too.

Maybe you have a different convention in xdsl. In that case, ignore my comment ;)

**Review by ingomueller-net — COMMENTED** [review]
Cool, thanks! LGTM!

**Pull request merged**